core/pool_vector.h

```diff
@@ -31,262 +31,290 @@
+#ifdef GODOT_POOL_VECTOR_USE_GLOBAL_LOCK
+			MemoryPool::global_mutex.unlock();
+#endif
+			unlock();
+		}
 
-		//copy the alloc data
-		alloc->size = old_alloc->size;
-		alloc->refcount.init();
-		alloc->pool_id = POOL_ALLOCATOR_INVALID_ID;
-		alloc->lock.set(0);
+		void lock(bool p_report_simultaneous = false) {
+#ifdef GODOT_POOL_VECTOR_USE_LOCK_PER_ALLOC
+			// Can uncomment this version to show when contention is occurring.
+			// if (mutex.try_lock() != OK) {
+			// WARN_PRINT("Contended poolvector lock.");
+			// mutex.lock();
+			// }
+			mutex.lock();
+#endif
 
-#ifdef DEBUG_ENABLED
-		MemoryPool::total_memory += alloc->size;
-		if (MemoryPool::total_memory > MemoryPool::max_memory) {
-			MemoryPool::max_memory = MemoryPool::total_memory;
+#ifdef GODOT_POOL_VECTOR_USE_CONTENTION_MONITOR
+			if ((contention_lock.postincrement() != 0) && p_report_simultaneous) {
+				WARN_PRINT_ONCE("Poolvector simultaneous access detected, possible race condition.");
+			}
+#endif
 		}
+		void unlock() {
+#ifdef GODOT_POOL_VECTOR_USE_LOCK_PER_ALLOC
+			mutex.unlock();
 #endif
 
-		MemoryPool::alloc_mutex.unlock();
+#ifdef GODOT_POOL_VECTOR_USE_CONTENTION_MONITOR
+			contention_lock.postdecrement();
+#endif
+		}
+	};
+
+	Alloc *_alloc = nullptr;
 
-		if (MemoryPool::memory_pool) {
-		} else {
-			alloc->mem = memalloc(alloc->size);
+	// Returns true if makes a copy.
+	bool _copy_on_write() {
+		if (!_alloc) {
+			return false;
 		}
 
-		{
-			Write w;
-			w._ref(alloc);
-			Read r;
-			r._ref(old_alloc);
-
-			int cur_elements = alloc->size / sizeof(T);
-			T *dst = (T *)w.ptr();
-			const T *src = (const T *)r.ptr();
-			for (int i = 0; i < cur_elements; i++) {
-				memnew_placement(&dst[i], T(src[i]));
-			}
+		// Note: This comment is preserved from the old implementation, possibly indicates some past problem with locking here.
+		// ERR_FAIL_COND(alloc->lock>0); should not be illegal to lock this for copy on write, as it's a copy on write after all
+
+		// Refcount should not be zero, otherwise it's a misuse of COW
+		uint32_t refcount = _alloc->refcount.get();
+		if (refcount == 1) {
+			return false; //nothing to do
 		}
+		DEV_ASSERT(refcount);
 
-		if (old_alloc->refcount.unref()) {
-			//this should never happen but..
+		Alloc *alloc = memnew(Alloc);
+		alloc->refcount.init(1);
+
+#ifdef GODOT_POOL_VECTOR_REPORT_LEAKS
+		MemoryPool::report_alloc(alloc, __LINE__);
+#endif
+
+		// copy contents
+		_alloc->super_lock();
+		alloc->vector = _alloc->vector;
+		_alloc->super_unlock();
 
 #ifdef DEBUG_ENABLED
-			MemoryPool::alloc_mutex.lock();
-			MemoryPool::total_memory -= old_alloc->size;
-			MemoryPool::alloc_mutex.unlock();
+		MemoryPool::counter_mutex.lock();
+		MemoryPool::allocs_used++;
+		MemoryPool::total_memory += (alloc->vector.size() * sizeof(T)) + sizeof(Alloc);
+		if (MemoryPool::total_memory > MemoryPool::max_memory) {
+			MemoryPool::max_memory = MemoryPool::total_memory;
+		}
+		MemoryPool::counter_mutex.unlock();
```

Comment:
I've already tried this and it works, but would prefer to put this in a separate PR, because the benefits of multiple `SafeNumeric`s versus a single `Mutex` need to be confirmed. (Presently it is no worse than previous version, as it works in the same way).